clang/docs/DPCPPToolsRFC.md

```diff
@@ -0,0 +1,260 @@
+responsible for initiating any of the link-time processing that occurs for each
+device target.
+
+An updated offloading model uses the new `clang-linker-wrapper` tool. Much of the
+functionality that is performed during the link phase of the offloading
+compilation is removed from the driver and moved to the `clang-linker-wrapper`
+tool.
+
+Below is a general representation of the overall offloading flow that is
+performed during a full compilation from source to final executable. The
+compiler driver is responsible for creating the multi-targeted object and the
+`clang-linker-wrapper` tool is responsible for the general functionality that is
+performed during the link. This compilation is capable of supporting multiple
+device targets. Support for Intel-based device targets relies on SPIR-V code
+generation. AMDGPU and NVPTX devices can be supported without relying on SPIR-V
+code generation, as shown in the figure below.
+
+![High level view of the offloading flow](images/offloadflow.svg)
+
+*Diagram 1: Overall compilation flow*
+
+## Multi-targeted object generation for SYCL offload kinds using `clang-offload-packager`
+
+`clang-offload-packager` plays a vital role during multi-targeted object generation. The
+multi-targeted object in the proposed offloading model is generated during the host
+compilation. The host compilation takes an additional argument which points to
+the device binary which will be embedded in the final object. Generation will
+be separated out to allow for potential parallelism during compilation of both
+the host and target device binaries.
+
+When dealing with multiple device binaries, an additional step is performed to
+package the multiple device binaries before being added to the host object.
+This additional step is performed with the `clang-offload-packager` taking image
+inputs containing information relating to the target triple, architecture
+setting and offloading kind.
+
+The `clang-offload-packager` is run during \'multi-targeted object\' generation regardless
+of the number of device binaries being added to the conglomerate multi-targeted object.
+The device binaries are contained in what is designated as an 'Offload Binary'.
+These binaries can reside in a variety of binary formats including Bitcode
+files, ELF objects, executables and shared objects, COFF objects, archives or
+simply stored as an offload binary.
+
+We should have the ability to package SPIR-V based device binaries in the
+offload section of any given binary. These device binaries will be packaged as
+normal with the packager and placed within the given section.
+
+Example usage of the external `clang-offload-packager` call:
+
+`clang-offload-packager --image=file=<name>,triple=<triple>,kind=<kind>`
+
+In the proposed offloading model, the compiler driver is responsible for
+creating the multi-targeted object. LLVM IR is used to represent device code
+embedded in this multi-targeted object. LLVM IR is known to change across
```

Comment:
@gmlueck 

Can you please take a look at the changes here and check if it is in lines of what we discussed?

Thanks

Reply:
I think this looks good, but a couple comments:

* Start a new paragraph with the sentence "LLVM IR is used to represent...".  This way, the whole paragraph is focused on this issue about LLVM IR instability.

* I'd say "LLVM IR is known to change across releases of clang ... ".  (Adding the word "clang" to clarify which releases we're talking about.)

* I wasn't sure what we meant by "and several other pre-compiled multi-targeted objects".  Would it make sense to say "... in static libraries and object files which contain device code IR"?

For the last two sentences, I wonder if we should say it like this:

> We are considering an alternative approach where static libraries and object files contain a device-specific IR instead of LLVM IR.  For Intel GPUs, we would use SPIR-V, which is more stable than LLVM IR.  We request comments on this alternate approach from the community.

This would head off concerns that SPIR-V might not be appropriate for other vendor's GPUs like CUDA.